compiler/rustc_mir/src/transform/instrument_coverage.rs

```diff
@@ -589,126 +640,212 @@ impl<'a, 'tcx> Instrumentor<'a, 'tcx> {
         op: Op,
         rhs: ExpressionOperandId,
     ) -> CoverageKind {
-        CoverageKind::Expression { id: self.next_expression(), lhs, op, rhs }
+        let id = self.next_expression();
+        CoverageKind::Expression { id, lhs, op, rhs }
     }
 
-    fn inject_statement(
-        &mut self,
-        file_name: Symbol,
-        source_file: &Lrc<SourceFile>,
-        coverage_kind: CoverageKind,
-        span: Span,
-        block: BasicBlock,
-    ) {
-        let code_region = make_code_region(file_name, source_file, span);
-        debug!("  injecting statement {:?} covering {:?}", coverage_kind, code_region);
-
-        let data = &mut self.mir_body[block];
-        let source_info = data.terminator().source_info;
-        let statement = Statement {
-            source_info,
-            kind: StatementKind::Coverage(box Coverage { kind: coverage_kind, code_region }),
-        };
-        data.statements.push(statement);
+    /// Counter IDs start from one and go up.
+    fn next_counter(&mut self) -> CounterValueReference {
+        assert!(self.next_counter_id < u32::MAX - self.num_expressions);
+        let next = self.next_counter_id;
+        self.next_counter_id += 1;
+        CounterValueReference::from(next)
     }
 
-    /// Converts the computed `BasicCoverageBlock`s into `SpanViewable`s.
-    fn span_viewables(&self, coverage_spans: &Vec<CoverageSpan>) -> Vec<SpanViewable> {
-        let tcx = self.tcx;
-        let mir_body = &self.mir_body;
-        let mut span_viewables = Vec::new();
-        for coverage_span in coverage_spans {
-            let bcb = self.bcb_from_coverage_span(coverage_span);
-            let CoverageSpan { span, bcb_leader_bb: bb, coverage_statements, .. } = coverage_span;
-            let id = bcb.id();
-            let mut sorted_coverage_statements = coverage_statements.clone();
-            sorted_coverage_statements.sort_unstable_by_key(|covstmt| match *covstmt {
-                CoverageStatement::Statement(bb, _, index) => (bb, index),
-                CoverageStatement::Terminator(bb, _) => (bb, usize::MAX),
-            });
-            let tooltip = sorted_coverage_statements
-                .iter()
-                .map(|covstmt| covstmt.format(tcx, mir_body))
-                .collect::<Vec<_>>()
-                .join("\n");
-            span_viewables.push(SpanViewable { bb: *bb, span: *span, id, tooltip });
-        }
-        span_viewables
+    /// Expression IDs start from u32::MAX and go down because a Expression can reference
+    /// (add or subtract counts) of both Counter regions and Expression regions. The counter
+    /// expression operand IDs must be unique across both types.
+    fn next_expression(&mut self) -> InjectedExpressionId {
+        assert!(self.next_counter_id < u32::MAX - self.num_expressions);
+        let next = u32::MAX - self.num_expressions;
+        self.num_expressions += 1;
+        InjectedExpressionId::from(next)
     }
+}
```

Comment:
```suggestion
}

```

Reply:
This is no longer structured this way. Probably accidentally joined when restructuring the commits in reverse.